### ISSUE TITLE
Still work for inception block?

### ISSUE THREAD

**Issue by twmht** [mbojarski/VisualBackProp#1]
![main-qimg-6833c13a23c33584be86a30bae210cf9](https://user-images.githubusercontent.com/1567200/31976356-ada0c5c6-b969-11e7-8e3e-f879a6fd00d9.png)

How to deal with the concatenation with convolutions?

**Comment by mbojarski** [comment]
There are few possibilities here:
1. Average all feature maps, and just scale it up to the size of the previous ones. It is easiest, but less accurate.
2. As you concat feature maps, you average them within groups. You should have then 4 average feature maps. You scale them up using deconv, accordingly to the conv/maxpool used to create them. The resulting ones should be of the same size and then you average them. 

For both cases the remaining part stays as in the original.

Please let me know if that helps.

**Comment by twmht** [comment]
![googlenet3](https://user-images.githubusercontent.com/1567200/32084837-30339096-bafe-11e7-82b2-d29ee2b1984f.png)

how about intermediate maxpool? Do I need to do unpool after deconv?

And how about intermidate upsample( e.g., conv->deconv->conv) ? the backward process would be deconv->conv->deconv?

**Comment by mbojarski** [comment]
If you do unpool then you can get more accurate results, but it will get more complicated. In this method we  do simplification by just scaling up using deconvolution with proper parameters. So you can combine two convolutions or convolution and maxpool into one deconvolution. For example:
1x1 conv + 3x3 max pool -> 3x3 deconv
3x3 conv + 3x3 conv -> 5x5 deconv

For the second question, yes, that would work, but again, you can combine it into one deconv with proper parameters (adjust filter size, stride, padding...).

Please let me know if that help.

**Comment by twmht** [comment]
@mbojarski 

I am implementing `visualBackProp` on some well-known networks.

Here is one more question about restNet shortcut.

![600x299xresnet png pagespeed ic m1j-vkbwpb](https://user-images.githubusercontent.com/1567200/32429905-c00a3b00-c307-11e7-9168-73028f4d6e9d.png)

the input of `relu` is added with shortcut, I am not sure how to inverse that, since it has been added and feed into relu.

in tensorflow, the output of each residual block would be  `tf.nn.relu(x + shorcut)`, how to deal with shortcut when doing deconvolution?

thanks for your help!

**Comment by mbojarski** [comment]
As we do not take weights into account in this method, the resnet block can be "replaced" with single convolution and you can ignore skip connection, i.e. if you have conv(3,3,1,1) followed by conv(3,3,2,2), you can replace it with one conv(5,5,2,2). This will give you reasonable results. Note that, the skip connection contributes to feature maps, so it will have effect on visualization.

**Comment by twmht** [comment]
@mbojarski 

>  if you have conv(3,3,1,1) followed by conv(3,3,2,2), you can replace it with one conv(5,5,2,2)

I don't understand, the format of conv you defined is (kernel_h, kernel_w, stride_h, stride_w)?

If the final feature map size is same for both single convolution and  residual block, then we can replace the residual block with the single convolution?

**Comment by mbojarski** [comment]
Yes, format is (kernel_h, kernel_w, stride_h, stride_w).

Yes. Also you need to make sure that you take into account kernel sizes and strides properly like in the given example.

If you doing it for various network, I would suggest to try making it universal, i.e. you need to make it for various layer types, and various parallel path. Then, it should be able to handle most of the networks.

**Comment by twmht** [comment]
@mbojarski 

How about batch normalization? Do I need to consider batch normalization when averaging feature maps?

**Comment by mbojarski** [comment]
No, you can just skip batchnorm layers.

**Comment by twmht** [comment]
@mbojarski 

I am facing nan problem when doing element-wise multiplication, the value from element-wise multiplication is too large to handle.

for example, one of the activation value (`avgUp` for some layer)  is `6.78244e+37`, and after element-wise multiplication, it may become too large to handle.

Do you have any advise to solve such case?

thanks

**Comment by twmht** [comment]
@mbojarski 

this problem solved after I replace each residual block with only one deconvolution.

However, if there are too many residual blocks (very deep network), the value may still be `nan`.

I think there are two ways to deal with the very deep network

one way is to replace several residual blocks with only one deconvolution, but the result may be less accurate, am I right?

the other way is to limit the maximum activation value after element-wise multiplication, for example, clip the maximum value to 255, but I am not sure if this is a good way.

**Comment by mbojarski** [comment]
This problem you can resolve by normalizing averaged feature max at each level. In particular, just divide averaged feature map by max of absolute values. This will prevent exploding and vanishing of values. Do not clip as it will change the result.

**Comment by getaleks** [comment]
> No, you can just skip batchnorm layers.

Why can batchnorm layers just be skipped? Doesn't this somehow influence the output? 

What should be done with kernel regularization in conv layers?

**Comment by mbojarski** [comment]
Norms are linear operations, and in the visualization algorithm you usually norm at the end, so you can skip norms in the middle.

**Comment by getaleks** [comment]
@mbojarski
I assume if you add Batch Normalization after the activation function, we need to skip it because of the shift parameters in the BN layer (usually called beta)? I guess the scaling factor (gamma) wouldn't be a big deal, but since the activations are being shifted VisualBackProp does not account for this?
